### PR TITLE
Issue #418: Allow disabling openapi documentation sites and schema download endpoints via config

### PR DESCRIPTION
**Pull request by ingjavierpinilla**
It solves [418](https://github.com/starlite-api/starlite/issues/418) and the subsequent suggestions made during the discussion

TLDR:
mainly but not limited to

- Add a list of enabled endpoints for openapi documentation.
- Change openapi controller return type from string to return objects.

# PR Checklist

- [x] Have you followed the guidelines in `CONTRIBUTING.md`?
- [x] Have you got 100% test coverage on new code?
- [x] Have you updated the prose documentation?
- [x] Have you updated the reference documentation?

### CODE REVIEW

**Review by cofin — CHANGES_REQUESTED** [review]
Thanks for the change here.  It's something I was thinking we needed as well.  Can you please update the documentation for configuration instructions before we merge this?

File: starlite/openapi/controller.py [hunk 1]
```diff
@@ -65,6 +67,57 @@ def get_schema_from_request(request: Request) -> "OpenAPI":
             raise ImproperlyConfiguredException("Starlite has not been instantiated with OpenAPIConfig")
         return request.app.openapi_schema
 
+    @staticmethod
+    def check_exposed_endpoints(request: Request) -> Optional[Response]:
```
Comment (peterschutt) [inline starlite/openapi/controller.py:71]: Sorry if it got a bit confusing in the issue discussion before, but we came to the conclusion that this needed to be a method that was called from within the route handlers to return the response, rather than via the `before_request` hook, so that the method could be easily overridden if desired.

This was also the reason that I suggested we change the response type of the handler methods so that they return a `Response` object in all cases, otherwise it would have been fine to leave those as `str` if we were going to use the `before_request` approach.

Pseudo code, but something along these lines:

```python
    @get(path="/elements", include_in_schema=False)
    def stoplight_elements(self, request: Request) -> Response:
        if should_serve_elements():
            # return the 200 response
        else:
            # return the 404 response
```
It'll work around those validation errors in CI too

Comment (ingjavierpinilla) [reply starlite/openapi/controller.py:71]: Alright, thanks for the clarification.
I'll change it o7

File: starlite/config/openapi.py [hunk 2]
```diff
@@ -107,6 +107,10 @@ class OpenAPIConfig(BaseModel):
     """
         The static schema generator to use for the "root" path of `/schema/`.
     """
+    allowed_endpoints: Set[str] = {"redoc", "swagger", "elements", "openapi.json", "openapi.yaml"}
```
Comment (Goldziher) [inline starlite/config/openapi.py:110]: I suggest to rename this to `enabled_endpoints` - it's a bit clearer

File: starlite/openapi/controller.py [hunk 3]
```diff
@@ -65,6 +66,31 @@ def get_schema_from_request(request: Request) -> "OpenAPI":
             raise ImproperlyConfiguredException("Starlite has not been instantiated with OpenAPIConfig")
         return request.app.openapi_schema
 
+    @staticmethod
+    def should_serve_endpoint(request: Request) -> bool:
+        """This method verifies that the requested path is within the allowed
```
Comment (Goldziher) [inline starlite/openapi/controller.py:71]: You are missing documentation for the args.

File: starlite/openapi/controller.py [hunk 4]
```diff
@@ -65,6 +66,31 @@ def get_schema_from_request(request: Request) -> "OpenAPI":
...
+            [ImproperlyConfiguredException][starlite.exceptions.ImproperlyConfiguredException]: If the application
+                `openapi_config` attribute is `None`.
+        """
+        if not request.app.openapi_config:  # pragma: no cover
+            raise ImproperlyConfiguredException("Starlite has not been instantiated with OpenAPIConfig")
+
+        path = set(filter(None, request.url.path.split("/")))
+
+        config = request.app.openapi_config
+        if path == {"schema"} and config.root_schema_site in config.allowed_endpoints:
```
Comment (Goldziher) [inline starlite/openapi/controller.py:87]: This is not going to work. A user can subclass the controller and change the root path, e.g. from `/schema` to `/api-docs`. Additionally the user can override the methods to change their paths.

Comment (ingjavierpinilla) [reply starlite/openapi/controller.py:87]: You're right, I will rethink how to validate when accessing the root, in case root_schema_site is not in enabled_endpoints (which should not happen, but I still think it is important to check it).

File: starlite/openapi/controller.py [hunk 5]
```diff
@@ -65,6 +66,35 @@ def get_schema_from_request(request: Request) -> "OpenAPI":
             raise ImproperlyConfiguredException("Starlite has not been instantiated with OpenAPIConfig")
         return request.app.openapi_schema
 
+    def should_serve_endpoint(self, request: Request) -> bool:
+        """This method verifies that the requested path is within the allowed
+        endpoints in the openapi_config.
+
+        Args:
+            request: A [Starlite][starlite.connection.Request] instance.
```
Comment (peterschutt) [inline starlite/openapi/controller.py:74]:
```suggestion
            request: To be tested if endpoint activated.
```
Referencing the type of parameters in the description is unnecessary as the ref docs framework already renders that information with the arg description.

image

File: starlite/openapi/controller.py [hunk 6]
```diff
@@ -65,6 +66,35 @@ def get_schema_from_request(request: Request) -> "OpenAPI":
...
+
+        Returns:
+            A boolean.
+
+        Raises:
+            [ImproperlyConfiguredException][starlite.exceptions.ImproperlyConfiguredException]: If the application
+                `openapi_config` attribute is `None`.
```
Comment (peterschutt) [inline starlite/openapi/controller.py:81]:
```suggestion
            ImproperlyConfiguredException: If the application `openapi_config` attribute is `None`.
```

No need to cross-ref the type, it will get resolved as a cross ref without it, e.g:
Before:

image

after

image

File: starlite/openapi/controller.py [hunk 7]
```diff
@@ -95,12 +129,25 @@ def retrieve_schema_yaml(self, request: Request) -> "OpenAPI":
                 A [Request][starlite.connection.Request] instance.
 
         Returns:
-            A rendered YAML object..
+            A [Response][starlite.response.Response] instance with a rendered YAML object
```
Comment (peterschutt) [inline starlite/openapi/controller.py:132]: Again, redundant to reference the type in the description:

image

**Review by peterschutt — CHANGES_REQUESTED** [review]
Quick review from a documentation perspective that highlights the workings of reference docs framework. I know some of these points are not consistent with existing docs on the controller, but that doesn't mean we shouldn't be moving forward with good practices! Good stuff, cheers:)

File: starlite/openapi/controller.py [hunk 8]
```diff
@@ -58,13 +59,40 @@ def get_schema_from_request(request: Request) -> "OpenAPI":
             An [OpenAPI][pydantic_openapi_schema.v3_1_0.open_api.OpenAPI] instance.
 
         Raises:
-            [ImproperlyConfiguredException][starlite.exceptions.ImproperlyConfiguredException]: If the application
-                `openapi_schema` attribute is `None`.
+            ImproperlyConfiguredException: If the application `openapi_config` attribute is `None`.
```
Comment (Goldziher) [inline starlite/openapi/controller.py:62]: Why this change? The reference docs wouldnt be generated like this. Unless there is a good reason for this, please change back.

File: starlite/openapi/controller.py [hunk 9]
```diff
@@ -58,13 +59,40 @@ def get_schema_from_request(request: Request) -> "OpenAPI":
...
         if not request.app.openapi_schema:  # pragma: no cover
             raise ImproperlyConfiguredException("Starlite has not been instantiated with OpenAPIConfig")
         return request.app.openapi_schema
 
+    def should_serve_endpoint(self, request: Request) -> bool:
+        """This method verifies that the requested path is within the enabled
+        endpoints in the openapi_config.
+
+        Args:
+            request: To be tested if endpoint enabled.
```
Comment (Goldziher) [inline starlite/openapi/controller.py:73]: So, since we are generating the reference documentation from the doc strings, we need to follow the `mkdocstrings` notation here. This means that this needs to be updated to include the linkage, e.g.

```suggestion
            request: The [Request][starlite.connection.Request] instance.
```

File: starlite/openapi/controller.py [hunk 10]
```diff
@@ -58,13 +59,40 @@ def get_schema_from_request(request: Request) -> "OpenAPI":
...
+
+        Returns:
+            A boolean.
+
+        Raises:
+            ImproperlyConfiguredException: If the application `openapi_config` attribute is `None`.
```
Comment (Goldziher) [inline starlite/openapi/controller.py:79]: Same as previously, for the linkage:

```suggestion
            [ImproperlyConfiguredException][starlite.exceptions.ImproperlyConfiguredException]: If the application `openapi_config` attribute is `None`.
```

File: starlite/openapi/controller.py [hunk 11]
```diff
@@ -58,13 +59,40 @@ def get_schema_from_request(request: Request) -> "OpenAPI":
...
+        """
+        if not request.app.openapi_config:  # pragma: no cover
+            raise ImproperlyConfiguredException("Starlite has not been instantiated with OpenAPIConfig")
```
Comment (Goldziher) [inline starlite/openapi/controller.py:82]:
```suggestion
            raise ImproperlyConfiguredException("Starlite has not been instantiated with an OpenAPIConfig")
```

File: starlite/openapi/controller.py [hunk 12]
```diff
@@ -58,13 +59,40 @@ def get_schema_from_request(request: Request) -> "OpenAPI":
...
+
+        request_path = set(filter(None, request.url.path.split("/")))
+        root_path = set(filter(None, self.path.split("/")))
+
+        config = request.app.openapi_config
+        if request_path == root_path and config.root_schema_site in config.enabled_endpoints:
+            return True
+
+        if request_path & config.enabled_endpoints:
```
Comment (Goldziher) [inline starlite/openapi/controller.py:91]: I suggest to `or` these two conditions instead of having two if clauses. Alternatively, save them into variables with clear names and then check `if <var1> or <var2>:`

File: starlite/openapi/controller.py [hunk 13]
```diff
@@ -95,12 +127,25 @@ def retrieve_schema_yaml(self, request: Request) -> "OpenAPI":
                 A [Request][starlite.connection.Request] instance.
 
         Returns:
-            A rendered YAML object..
+            resoponse: With a rendered YAML object
```
Comment (Goldziher) [inline starlite/openapi/controller.py:130]: `resoponse`.

Also, here too we need linkage, thus:

```suggestion
            A [Response][starlite.response.Response] instance with the YAML object rendered into a string.
```

File: starlite/openapi/controller.py [hunk 14]
```diff
@@ -109,12 +154,25 @@ def retrieve_schema_json(self, request: Request) -> "OpenAPI":
                 A [Request][starlite.connection.Request] instance.
 
         Returns:
-            A rendered JSON object..
+            resoponse: With a rendered YAML object
```
Comment (Goldziher) [inline starlite/openapi/controller.py:157]: Multiple issues here - both a typo and `YAML` instead of `JSON`.

```suggestion
            A [Response][starlite.response.Response] instance with the JSON object rendered into a string.
```

File: starlite/openapi/controller.py [hunk 15]
```diff
@@ -95,12 +127,25 @@ def retrieve_schema_yaml(self, request: Request) -> "OpenAPI":
...
+        if not request.app.openapi_config:  # pragma: no cover
+            raise ImproperlyConfiguredException("Starlite has not been instantiated with OpenAPIConfig")
+
+        if self.should_serve_endpoint(request):
+            return Response(
+                content=self.get_schema_from_request(request),
+                status_code=HTTP_200_OK,
+                media_type=OpenAPIMediaType.OPENAPI_YAML,
+            )
+        return Response(
```
Comment (Goldziher) [inline starlite/openapi/controller.py:141]: ok, here the response should not be of type HTML. It should be of type `OpenAPIMediaType.OPENAPI_YAML` with content being an empty dictionary.

File: starlite/openapi/controller.py [hunk 16]
```diff
@@ -109,12 +154,25 @@ def retrieve_schema_json(self, request: Request) -> "OpenAPI":
...
+        if not request.app.openapi_config:  # pragma: no cover
+            raise ImproperlyConfiguredException("Starlite has not been instantiated with OpenAPIConfig")
+
+        if self.should_serve_endpoint(request):
+            return Response(
+                content=self.get_schema_from_request(request),
+                status_code=HTTP_200_OK,
+                media_type=OpenAPIMediaType.OPENAPI_JSON,
+            )
+        return Response(
```
Comment (Goldziher) [inline starlite/openapi/controller.py:168]: same comment basically as above regarding this response not being of the correct type. This should be a `OpenAPIMediaType.OPENAPI_JSON` response with an empty dictionary.

File: starlite/openapi/controller.py [hunk 17]
```diff
@@ -125,56 +183,83 @@ def root(self, request: Request) -> str:
                 A [Request][starlite.connection.Request] instance.
 
         Returns:
-            A rendered html string.
+            resoponse: With the rendered site defined in root_schema_site.
```
Comment (Goldziher) [inline starlite/openapi/controller.py:186]: `resoponse` 😉 , you should be more careful with copy/paste.

Please link the response references here as above.

File: starlite/openapi/controller.py [hunk 18]
```diff
@@ -125,56 +183,83 @@ def root(self, request: Request) -> str:
...
 
         Raises:
-            [ImproperlyConfiguredException][starlite.exceptions.ImproperlyConfiguredException]: If the application
-                `openapi_config` attribute is `None`.
+            ImproperlyConfiguredException: If the application `openapi_config` attribute is `None`.
```
Comment (Goldziher) [inline starlite/openapi/controller.py:189]: missing linkage, please revert this change.

File: starlite/openapi/controller.py [hunk 19]
```diff
@@ -125,56 +183,83 @@ def root(self, request: Request) -> str:
...
+    def swagger_ui(self, request: Request) -> Response:
         """Route handler responsible for rendering Swagger-UI.
 
         Args:
             request:
                 A [Request][starlite.connection.Request] instance.
 
         Returns:
-            A rendered html string.
+            response: With a rendered swagger documentation site
```
Comment (Goldziher) [inline starlite/openapi/controller.py:213]: No typo! but missing linkage, please update.

File: starlite/openapi/controller.py [hunk 20]
```diff
@@ -125,56 +183,83 @@ def root(self, request: Request) -> str:
...
+    def stoplight_elements(self, request: Request) -> Response:
         """Route handler responsible for rendering StopLight Elements.
 
         Args:
             request:
                 A [Request][starlite.connection.Request] instance.
 
         Returns:
-            A rendered html string.
+            response: With a rendered stoplight elements documentation site
```
Comment (Goldziher) [inline starlite/openapi/controller.py:235]: same as prev.

**Review by Goldziher — CHANGES_REQUESTED** [review]
Thanks for the PR!

Some issues - primarily with the documentation. Please address these. You can then run

```shell
docker compose up
```

To launch the doc site locally and check all is well. Once the doc issues are addressed, I think this can be merged.

File: starlite/openapi/controller.py (hunk 8 above)
Comment (peterschutt) [reply starlite/openapi/controller.py:62]: @Goldziher I think you should check out my previous review

Comment (Goldziher) [reply starlite/openapi/controller.py:62]: ok, as the reference tzar decides 😉

File: starlite/openapi/controller.py (hunk 9 above)
Comment (peterschutt) [reply starlite/openapi/controller.py:73]: These too, I suggested we don't need to cross-ref the type of the arg in the description as the docs already work it out:

image

File: starlite/openapi/controller.py (hunk 12 above)
Comment (ingjavierpinilla) [reply starlite/openapi/controller.py:91]: I'm not a big fan of this idea, I prefer separated and simple if statements. It's easier to read and understand imho

Comment (Goldziher) [reply starlite/openapi/controller.py:91]: as you see fit

**Comment by Goldziher** [comment]
@all-contributors add @ingjavierpinilla code
